packages/contracts/contracts/plugin/PluginRepo.sol

```diff
@@ -0,0 +1,233 @@
+    /// @return pluginFactoryAddress Address of latest plugin factory for version
+    /// @return contentURI External URI for fetching latest version's content
+    function getLatest()
+        public
+        view
+        returns (
+            uint16[3] memory semanticVersion,
+            address pluginFactoryAddress,
+            bytes memory contentURI
+        )
+    {
+        return getByVersionId(nextVersionIndex - 1);
+    }
+
+    /// @notice get latest by plugin factory address
+    /// @return semanticVersion Semantic version for pluginRepo version
+    /// @return pluginFactoryAddress Address of plugin factory for version
+    /// @return contentURI External URI for fetching version's content
+    function getLatestForContractAddress(address _pluginFactoryAddress)
+        public
+        view
+        returns (
+            uint16[3] memory semanticVersion,
+            address pluginFactoryAddress,
+            bytes memory contentURI
+        )
+    {
+        return getByVersionId(versionIdxForContract[_pluginFactoryAddress]);
+    }
+
+    /// @notice get latest by semantic version
+    /// @return semanticVersion Semantic version for latest pluginRepo version
+    /// @return pluginFactoryAddress Address of plugin factory for version
+    /// @return contentURI External URI for fetching latest version's content
+    function getBySemanticVersion(uint16[3] memory _semanticVersion)
+        public
+        view
+        returns (
+            uint16[3] memory semanticVersion,
+            address pluginFactoryAddress,
+            bytes memory contentURI
+        )
+    {
+        return getByVersionId(versionIdxForSemantic[semanticVersionHash(_semanticVersion)]);
+    }
+
+    /// @notice get latest by version id
+    /// @return semanticVersion Semantic version for pluginRepo version
+    /// @return pluginFactoryAddress Address of plugin factory for version
+    /// @return contentURI External URI for fetching version's content
+    function getByVersionId(uint256 _versionIdx)
+        public
+        view
+        returns (
+            uint16[3] memory semanticVersion,
+            address pluginFactoryAddress,
+            bytes memory contentURI
+        )
+    {
+        if (_versionIdx <= 0 || _versionIdx >= nextVersionIndex) revert VersionDoesNotExist();
+        Version storage version = versions[_versionIdx];
+        return (version.semanticVersion, version.pluginFactoryAddress, version.contentURI);
+    }
+
+    /// @notice get version count
+    /// @return uint256 count value
+    function getVersionCount() public view returns (uint256) {
+        return nextVersionIndex - 1;
+    }
+
+    /// @notice check if new version is valid compared to the previous one
+    /// @param _oldVersion Old semantic version
+    /// @param _newVersion New semantic version
+    /// @return bool True if bump is valid
+    function isValidBump(uint16[3] memory _oldVersion, uint16[3] memory _newVersion)
```

Comment:
not sure about this:
1. this doesn't allow a dev to maintain 2 different major versions. if I understand it correctly I cannot add a patch to version 1 when I already deployed version 2
2. line 213 forces a dev to follow semver properly. so it doesn't allow to skip a version (e.g. I cannot jump from 1.1.1 to 1.2.1)

Personally I would remove most or all of the logic because it isn't our responsibility

Reply:
task: https://aragonassociation.atlassian.net/browse/APP-635